Group all versioning concerns under a `VersionContext` umbrella

tomwadeson opened the pull request:
Having a play with application wiring and dependency passing -- see what you think.

Commit:
Introduce a `VersionContext` to group all versioning dependencies

... and prefer explicit parameter passing.

Commit:
Tidy up examples

stettix commented on examples/src/test/scala/com/gu/tableversions/examples/DatePartitionedTableLoaderSpec.scala:
```diff
@@ -170,3 +170,13 @@ object DatePartitionedTableLoaderSpec {
+      metastore = new SparkHiveMetastore[IO]()
+      versionGenerator = Version.generateVersion
+    } yield VersionContext(tableVersions, metastore, versionGenerator)
+}
```
❤️

stettix commented on spark/src/main/scala/com/gu/tableversions/spark/VersionContext.scala:
```diff
@@ -17,46 +17,32 @@ import org.apache.spark.sql.{Dataset, Row, SaveMode, SparkSession}
+    tableVersions: TableVersions[IO],
+    metastore: Metastore[IO],
+    generateVersion: IO[Version]
+) {
```
This is a great change, just one thought: I wonder if it would make sense to move this `VersionContext` class into `core`? The `versionedInsertDatasetIntoTable` would then have to move out of it again, of course.

The reason I'm wondering about this is that this could make a good place to hang some functionality that interacts with both the `TableVersions` and the `Metastore`. For example the `TableLoader.checkout()` method: https://github.com/guardian/table-versions/blob/master/examples/src/main/scala/com/gu/tableversions/examples/TableLoader.scala#L47. This currently lives in example code, but it would be useful to have this somewhere in the real code. It doesn't belong in either of the two underlying services as it interacts with both. It also doesn't rely on Spark or any of the concrete implementations, so `core` would seem like a good place for it.

Likewise, the "commit and sync metastore" parts of the current `versionedInsertDatasetIntoTable` method could perhaps be pulled into common code int he same place.

What do you think?

tomwadeson replied:
Yep, that sounds like a better place for it. Would you like to jump in and move things around? I think you'll have a better sense for exactly where to locate things.

It would be nice to preserve the ergonomics of enriching `Dataset`/`Dataframe` with a versioned insert method *without* having to pass the `VersionContext` around implicitly. I think I've got an idea brewing for this, but would need to play with it.

stettix replied:
OK, I can have a quick go and see what it looks like. (Looking forward to hearing about your cunning idea!)

stettix replied:
So I had a play but without much luck. One issue is that the `Metastore` trait is in the separate `metastore` module; thinking about it, I think this module should be removed and the code go in `core`, as these are implementation-neutral interfaces and types, just as the `TableVersions` trait and related code.

Also, the `VersionContext` you've created here is really a Spark version context, which makes a lot of sense, collecting the bits needed to write Spark code in an easy to use package. That's probably a separate concept from any lower level service that doesn't know about Spark, and is useful even if such a thing is added. So I think the changes here are fine as they are.

stettix commented on spark/src/main/scala/com/gu/tableversions/spark/VersionContext.scala:
```diff
@@ -87,6 +73,27 @@ object VersionedDataset {
+  }
+}
+
+object VersionContext {
```
As a slight aside, I think there are some more `implicit`s you can get rid of in this code if you want, on the `partitionValues` and `writeVersionedPartitions` method?

stettix commented on examples/src/test/scala/com/gu/tableversions/examples/DatePartitionedTableLoaderSpec.scala:
```diff
@@ -171,3 +171,13 @@ object DatePartitionedTableLoaderSpec {
+      metastore = new SparkHiveMetastore[IO]()
+      versionGenerator = Version.generateVersion
+    } yield VersionContext(tableVersions, metastore, versionGenerator)
+}
```
BTW maybe pull out into its own file as it's used in a couple of tests?

stettix commented on examples/src/main/scala/com/gu/tableversions/examples/TableLoader.scala:
```diff
@@ -6,7 +6,7 @@ import cats.effect.IO
 import com.gu.tableversions.core.{TableDefinition, TableVersions, Version}
 import com.gu.tableversions.metastore.Metastore
-import com.gu.tableversions.spark.VersionedDataset
+import com.gu.tableversions.spark.VersionContext
```
I think you can remove some `import`s here now.

tomwadeson replied:
Who needs `-Ywarn-unused:imports` with eyes like that? 😀

stettix commented on spark/src/test/scala/com/gu/tableversions/spark/VersionContextSpec.scala:
```diff
@@ -105,19 +105,25 @@ class VersionedDatasetSpec extends FlatSpec with Matchers with SparkHiveSuite {
+      val tableVersions: TableVersions[IO] = (for {
+        t <- InMemoryTableVersions[IO]
+        _ <- t.init(usersTable.name, isSnapshot = true, UserId("test"), UpdateMessage("init"), Instant.now())
+      } yield t).unsafeRunSync()
```
This isn't related to your changes, but this bit of code is repeated a few times. Might be nice pulling it out somewhere. Could `TestVersionContext` be used here as well maybe? It would have to move into the `spark` module to do that, of course. But maybe that would be a step too far...

tomwadeson replied:
I did have a look at this, but nothing jumped out quickly.

There's a bit of me that quite likes it anyway - makes it easy to see the behaviour that we're going to observe when interacting with the `VersionContext` in the test.

Will keep an eye on it!

Commit:
Extract a `SparkSupport` module from `VersionContext`

This removes any awareness of Spark from `VersionContext`, potentially allowing for it
to be reused in projects without that dependency.

Commit:
Remove unused imports and params

Commit:
Locate `TestVersionContext` in its own source file

stettix reviewed (approved):
Nice work! 👍
